test/Hessian_test.cpp

```diff
@@ -0,0 +1,433 @@
+
+    std::vector<AD> x = {3.0, 2.0};
+
+    // f(x) = x[0]^2 + x[1]^2
+    auto foo = [](std::vector<AD> &x) -> AD { return x[0] * x[0] + x[1] * x[1]; };
+
+    std::vector<std::vector<AD>> expected_hessian = {{2.0, 0.0}, {0.0, 2.0}};
+
+    auto computed_hessian = xad::computeHessian<double>(x, foo, &tape);
+
+    for (unsigned int i = 0; i < expected_hessian.size(); i++)
+        for (unsigned int j = 0; j < expected_hessian[i].size(); j++)
+            ASSERT_EQ(expected_hessian[i][j], computed_hessian[i][j]);
+}
+
+TEST(HessianTest, QuadraticForwardAdjointAutoTape)
+{
+    typedef xad::fwd_adj<double> mode;
+    typedef mode::active_type AD;
+
+    std::vector<AD> x = {3.0, 2.0};
+
+    // f(x) = x[0]^2 + x[1]^2
+    auto foo = [](std::vector<AD> &x) -> AD { return x[0] * x[0] + x[1] * x[1]; };
+
+    std::vector<std::vector<AD>> expected_hessian = {{2.0, 0.0}, {0.0, 2.0}};
+
+    auto computed_hessian = xad::computeHessian<double>(x, foo);
+
+    for (unsigned int i = 0; i < expected_hessian.size(); i++)
+        for (unsigned int j = 0; j < expected_hessian[i].size(); j++)
+            ASSERT_EQ(expected_hessian[i][j], computed_hessian[i][j]);
+}
+
+TEST(HessianTest, QuadraticForwardAdjointFetchTape)
+{
+    typedef xad::fwd_adj<double> mode;
+    typedef mode::active_type AD;
+    typedef mode::tape_type tape_type;
+
+    tape_type tape;
+
+    std::vector<AD> x = {3.0, 2.0};
+
+    // f(x) = x[0]^2 + x[1]^2
+    auto foo = [](std::vector<AD> &x) -> AD { return x[0] * x[0] + x[1] * x[1]; };
+
+    std::vector<std::vector<AD>> expected_hessian = {{2.0, 0.0}, {0.0, 2.0}};
+
+    auto computed_hessian = xad::computeHessian<double>(x, foo);
+
+    for (unsigned int i = 0; i < expected_hessian.size(); i++)
+        for (unsigned int j = 0; j < expected_hessian[i].size(); j++)
+            ASSERT_EQ(expected_hessian[i][j], computed_hessian[i][j]);
+}
+
+TEST(HessianTest, QuadraticForwardAdjointWithIterator)
+{
+    typedef xad::fwd_adj<double> mode;
+    typedef mode::active_type AD;
+    typedef mode::tape_type tape_type;
+
+    tape_type tape;
+
+    std::vector<AD> x = {3.0, 2.0};
+
+    // f(x) = x[0]^2 + x[1]^2
+    auto foo = [](std::vector<AD> &x) -> AD { return x[0] * x[0] + x[1] * x[1]; };
+
+    std::list<std::list<AD>> expected_hessian = {{2.0, 0.0}, {0.0, 2.0}};
+
+    std::list<std::list<AD>> computed_hessian(x.size(), std::list<AD>(x.size(), 0.0));
+    xad::computeHessian<decltype(begin(computed_hessian)), double>(x, foo, begin(computed_hessian),
+                                                                   end(computed_hessian), &tape);
+
+    auto row1 = computed_hessian.begin(), row2 = expected_hessian.begin();
+    while (row1 != computed_hessian.end() && row2 != expected_hessian.end())
+    {
+        auto col1 = row1->begin(), col2 = row2->begin();
+        while (col1 != row1->end() && col2 != row2->end())
+        {
+            ASSERT_EQ(*col1, *col2);
+            col1++;
+            col2++;
+        }
```

Comment:
That `PointWise(DoubleEq(), ...)` stuff from gmock works also with lists, so it would be simpler...